src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp

```diff
@@ -0,0 +1,264 @@
+        return ReturnCode_t::RETCODE_OK;
+    }
+
+    ReturnCode_t return_loan(
+            void* sample)
+    {
+        OutstandingLoanItem* item = find_by_sample(sample);
+
+        if (nullptr == item)
+        {
+            return ReturnCode_t::RETCODE_BAD_PARAMETER;
+        }
+
+        item->num_refs -= 1;
+        if (item->num_refs == 0)
+        {
+            CacheChange_t tmp;
+            tmp.payload_owner(item->owner);
+            tmp.serializedPayload = item->payload;
+            item->owner->release_payload(tmp);
+            item->payload.data = nullptr;
+            item->owner = nullptr;
+
+            item = free_loans_.push_back(*item);
+            assert(nullptr != item);
+            used_loans_.remove(*item);
+        }
+
+        return ReturnCode_t::RETCODE_OK;
+    }
+
+private:
+
+    struct OutstandingLoanItem
+    {
+        void* sample = nullptr;
+        SampleIdentity identity;
+        SerializedPayload_t payload;
+        IPayloadPool* owner = nullptr;
+        uint32_t num_refs = 0;
+
+        ~OutstandingLoanItem()
+        {
+            payload.data = nullptr;
+        }
+
+        OutstandingLoanItem() = default;
+        OutstandingLoanItem(
+                const OutstandingLoanItem&) = default;
+        OutstandingLoanItem& operator =(
+                const OutstandingLoanItem&) = default;
+        OutstandingLoanItem(
+                OutstandingLoanItem&&) = default;
+        OutstandingLoanItem& operator =(
+                OutstandingLoanItem&&) = default;
+
+        bool operator == (
+                const OutstandingLoanItem& other) const
+        {
+            return other.sample == sample && other.payload.data == payload.data;
+        }
+
+    };
+
+    using collection_type = eprosima::fastrtps::ResourceLimitedVector<OutstandingLoanItem>;
+
+    eprosima::fastrtps::ResourceLimitedContainerConfig limits_;
+    collection_type free_loans_;
+    collection_type used_loans_;
+    TypeSupport type_;
+
+    OutstandingLoanItem* find_by_change(
+            CacheChange_t* change)
+    {
+        SampleIdentity id;
+        id.writer_guid(change->writerGUID);
+        id.sequence_number(change->sequenceNumber);
+
+        auto comp = [id](const OutstandingLoanItem& item)
+                {
+                    return id == item.identity;
+                };
+        auto it = std::find_if(used_loans_.begin(), used_loans_.end(), comp);
+        if (it != used_loans_.end())
+        {
+            return &(*it);
+        }
+        return nullptr;
+    }
+
+    OutstandingLoanItem* find_by_sample(
+            void* sample)
+    {
+        auto comp = [sample](const OutstandingLoanItem& item)
+                {
+                    return sample == item.sample;
+                };
+        auto it = std::find_if(used_loans_.begin(), used_loans_.end(), comp);
+        if (it != used_loans_.end())
+        {
+            return &(*it);
+        }
+        return nullptr;
```

Comment:
Never reached in tests.

Reply:
I don't think it is possible from the user API to make this happen, so I will turn the condition on line 250 into an assertion.

Reply:
Done in 6addbc6